examples/greeting/public/plugin.ts

```diff
@@ -0,0 +1,118 @@
+
+const defaultGreetingProvider: GreetingProvider = (def: GreetingDefinition) => ({
+  greetMe: (name: string) => alert(`${def.salutation} ${name}${def.punctuation}`),
+  label: def.id,
+});
+
+export interface GreetingStart {
+  /**
+   * This function should be used if the value of `id` is not known at compile time. Usually
+   * this is because `id` has been persisted somewhere. If the value of `id` is known at
+   * compile time (e.g. `greeting.getGreeter(CASUAL_GREETER))`) developers should prefer accessing
+   * off the plugin that registered `CasualGreeter`, like `customGreetings.getCasualGreeter()`, because:
+   *  - makes it more explicit you should add `customGreetings` to your plugin dependency list.
+   *  - don't need to handle the possibility of `undefined` if it's a required dependency.
+   *  - can get more specialized types in certain situations (e.g. `interface CustomGreeter` vs `Greeter` -
+   * does not apply to this example but many real world examples are like this.)
+   */
+  getGreeter: (id: string) => Greeter;
+  /**
+   * Returns an array of all registered greeters.
+   */
+  getGreeters: () => Greeter[];
+}
+
+export interface GreetingSetup {
+  /**
+   * Allows for another plugin to add a custom provider. Only one plugin is allowed to set a custom
+   * provider, any secondary attempts will result in an error.
+   */
+  setCustomProvider: (customProvider: GreetingProvider) => void;
+
+  /**
+   * Returns a greeting accessor that should only be used after the setup lifecycle completes.
+   */
+  registerGreetingDefinition: (greetingDefinition: GreetingDefinition) => () => Greeter;
+}
+
+export class GreetingPlugin implements Plugin<GreetingSetup, GreetingStart> {
+  private greetingDefinitions: { [key: string]: GreetingDefinition } = {};
+  /**
+   * Greeting provider is set at the beginning of this plugins start lifecycle to ensure
+   * any other plugins had a chance to set a custom provider.
+   */
+  private greetingProvider?: GreetingProvider;
+
+  /**
+   * Store a custom provider set by another plugin.
+   */
+  private customGreetingProvider?: GreetingProvider;
+
+  setup = () => ({
+    /**
+     * TODO: this only allows for a single greeting provider to be set. It would be nice
+     * to come up with a pattern that allowed for multiple custom providers to be registered.
+     */
+    setCustomProvider: (customProvider: GreetingProvider) => {
+      if (this.greetingProvider) {
+        throw new Error(
+          'Only one custom greeting provider is allowed, and one has already been registered.'
+        );
+      } else {
+        this.customGreetingProvider = customProvider;
+      }
+    },
+    registerGreetingDefinition: (greetingDefinition: GreetingDefinition) => {
+      this.greetingDefinitions[greetingDefinition.id] = greetingDefinition;
+      return () => {
+        if (!this.greetingProvider) {
+          throw new Error('Greeters can only be retrieved after setup lifecycle.');
```

Comment:
I think throwing an explicit error here is definitely a step in the right direction. There's still the potential for a run-time exception. However, this already exists with the way that the NP setup/start lifecycles work in practicality after `getStartServices` was introduced...